guides/micronaut-cache-hazelcast/micronaut-cache-hazelcast.adoc

```diff
@@ -0,0 +1,171 @@
+<6> Method invocation causes the invalidation of the cache `headlines` for the supplied `month`.
+
+TIP: If you don't annotate the class with `@CacheConfig`, specify the cache name in the cache annotations. E.g. `@Cacheable(value = "headlines", parameters = {"month"})`
+
+=== Test the Cache
+
+We use https://testcontainers.org[Testcontainers] to test the application.
+
+Create a class to ease the start-up and shutdown of Hazelcast with TestContainers.
+
+test:HazelcastUtils[]
+
+We can verify that the cache works as expected:
+
+test:NewsServiceTest[]
+
+:exclude-for-languages:groovy
+callout:test-method-order[1]
+callout:micronaut-test-start-application-false[2]
+callout:test-instance-per-class[3]
+callout:test-property-provider[4]
+<5> Inject `NewsService` bean.
+callout:timeout[6]
+callout:order[7]
+:exclude-for-languages:
+
+:exclude-for-languages:java,kotlin
+callout:stepwise[1]
+callout:micronaut-test-start-application-false[2]
+callout:test-property-provider[3]
+<4> Inject `NewsService` bean.
+callout:timeout[5]
+:exclude-for-languages:
+
+=== Controller
+
+Create a controller which engages the previous service:
+
+source:News[]
+
+callout:introspected[1]
+
+source:NewsController[]
+
+callout:controller[number=1,arg0=/]
+
+Add a test. We call the endpoint twice and verify with the `@Timeout` annotation that the cache is being used.
+
+test:NewsControllerTest[]
+
+:exclude-for-languages:groovy
+callout:test-method-order[1]
+callout:micronaut-test[2]
+callout:test-instance-per-class[3]
+callout:test-property-provider[4]
+callout:http-client[5]
+callout:timeout[6]
+callout:order[7]
+:exclude-for-languages:
+
+:exclude-for-languages:java,kotlin
+callout:stepwise[1]
+callout:micronaut-test[2]
+callout:test-property-provider[3]
+callout:http-client[4]
+callout:timeout[5]
+:exclude-for-languages:
+
+include::{commondir}/common-testApp.adoc[]
+
+=== Configure Hazelcast 
+
+The minimal configuration to use Hazelcast is to simply declare `hazelcast:` with a network configuration for addresses of the Hazelcast cluster:
+
+resource:application.yml[tag=hazelcast]
+
+Replace `192.168.4.67` with your host machine address. See following section.
+
+== Running Hazelcast via Docker 
+
+Run https://hub.docker.com/r/hazelcast/hazelcast[Hazelcast via Docker]
+
+Create a Hazelcast cluster with two Docker containers and access it from the client application.
+
+[source, bash]
+----
+$ docker run -e HZ_NETWORK_PUBLICADDRESS=192.168.4.67:5701 -p 5701:5701 hazelcast/hazelcast:4.2.1
+$ docker run -e HZ_NETWORK_PUBLICADDRESS=192.168.4.67:5702 -p 5702:5701 hazelcast/hazelcast:4.2.1
+----
+
+Replace `192.168.4.67` with your host machine address. It is used for the Hazelcast communication.
+
+include::{commondir}/common-runapp.adoc[]
+
+You should be able to execute `curl localhost:8080/NOVEMBER` and see results
+
+include::{commondir}/common-graal-with-plugins.adoc[]
```

Comment:
We need to remove this and any other reference to GraalVM in the guide (like in the guide metadata) because Hazelcast is not compatible with GraalVM.
I've tried with the agent to generate the appropriate reflection config and I "fix" a couple of issues, but now it fails with:

```
Caused by: java.io.InvalidClassException: java.util.Arrays$ArrayList; no valid constructor
	at java.io.ObjectStreamClass$ExceptionInfo.newInvalidClassException(ObjectStreamClass.java:159)
	at java.io.ObjectStreamClass.checkDeserialize(ObjectStreamClass.java:875)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2170)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1679)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:493)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:451)
	at com.hazelcast.internal.serialization.impl.defaultserializers.JavaDefaultSerializers$JavaSerializer.read(JavaDefaultSerializers.java:86)
	at com.hazelcast.internal.serialization.impl.defaultserializers.JavaDefaultSerializers$JavaSerializer.read(JavaDefaultSerializers.java:79)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:44)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:208)
```

Reply:
@nfrankel  do you have any insights about this ?

Reply:
> We need to remove this and any other reference to GraalVM in the guide (like in the guide metadata) because Hazelcast is not compatible with GraalVM.

It's possible to run Hazelcast via GraalVM, but obviously, you need to fine-tune the configuration according to your needs. Here's a toy project to demo it: https://github.com/nfrankel/hazelcast-native

Reply:
@nfrankel are there plans to ship that native image configuration on the hazelcast end ?

Reply:
Thanks for the example Nicolas, I'll give it a try again because maybe I missed some config from the agent.

Reply:
Short answer, no.

Long answer: the problem is that it depends on how you use Hazelcast. Every usage will be different, sometimes with custom code dynamically loaded. It makes it impossible to write a catch-all configuration.

The project above shows how to do it for one simple use-case, so I hope it might help people.

Reply:
@ilopmar let me know if it works and we can add some default configuration to the Micronaut Hazelcast cache dependency to allow GraalVM Native Images for at least the simplest applications.

Reply:
@sdelamo I have a minimal GraalVM config that works **for this application**. I've kept all the classes regarding Hazelcast in `reflect-config.json` but the tricky part is the `serialization-config.json` file. For our application I needed to include:
```
[
  {
    "name":"java.util.Arrays$ArrayList"
  },
  {
    "name":"java.lang.String"
  }
]
```

Because we are using those two classes in `NewsService` but that is something that users need to do for their application.

I've also added the config the bridge JUL (used by Hazelcast) to SLF4J so it's log is displayed properly by logback. It's the same I did here https://micronaut-projects.github.io/micronaut-liquibase/latest/guide/#_logging

I'm going to send a PR to micronaut-cache to include the `reflect-config.json` config and some docs about GraalVM and the logging configuration. I think we can add some experimental support for Hazelcast.

Reply:
I think we will publish the guide without GraalVM Section for now and we can add a GraalVM section once that it is in the cache module.